pkg/discovery/clustersetip/clustersetip.go

```diff
@@ -0,0 +1,234 @@
+	Enabled bool
+	cidr.Info
+}
+
+type Config struct {
+	ClusterID        string
+	ClustersetIPCIDR string
+	ClusterSize      uint
+}
+
+func AllocateClustersetIPCIDR(clustersetIPInfo *Info) (string, error) {
+	return cidr.Allocate(&clustersetIPInfo.Info) //nolint:wrapcheck // No need to wrap
+}
+
+func isCIDRPreConfigured(clusterID string, clustersetIPNetworks map[string]*cidr.ClusterInfo) bool {
+	// ClustersetIPCIDR is not pre-configured
+	if clustersetIPNetworks[clusterID] == nil || clustersetIPNetworks[clusterID].CIDRs == nil ||
+		len(clustersetIPNetworks[clusterID].CIDRs) == 0 {
+		return false
+	}
+
+	// ClustersetIPCIDR is pre-configured
+	return true
+}
+
+func ValidateClustersetIPConfiguration(clustersetIPInfo *Info, netconfig Config, status reporter.Interface) (string, error) {
+	status.Start("Validating ClustersetIP configuration")
+	defer status.End()
+
+	clustersetIPClusterSize := netconfig.ClusterSize
+	clustersetIPCIDR := netconfig.ClustersetIPCIDR
+
+	if clustersetIPClusterSize != 0 && clustersetIPClusterSize != clustersetIPInfo.AllocationSize {
+		clusterSize, err := cidr.GetValidClusterSize(clustersetIPInfo.CIDR, clustersetIPClusterSize)
+		if err != nil {
+			return "", status.Error(err, "invalid cluster size")
+		}
+
+		clustersetIPInfo.AllocationSize = clusterSize
+	}
+
+	if clustersetIPCIDR != "" && clustersetIPClusterSize != 0 {
+		status.Failure("Only one of cluster size and clustersetip CIDR can be specified")
+
+		return "", errors.New("only one of cluster size and clustersetip CIDR can be specified")
+	}
+
+	if clustersetIPCIDR != "" {
+		err := cidr.IsValid(clustersetIPCIDR)
+		if err != nil {
+			return "", errors.Wrap(err, "specified clustersetip-cidr is invalid")
+		}
+	}
+
+	return clustersetIPCIDR, nil
+}
+
+func GetClustersetIPNetworks(ctx context.Context, client controllerClient.Client, brokerNamespace string) (*Info, *v1.ConfigMap, error) {
+	configMap, err := GetConfigMap(ctx, client, brokerNamespace)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error retrieving clustersetip ConfigMap")
+	}
+
+	clustersetIPInfo := Info{}
+
+	err = json.Unmarshal([]byte(configMap.Data[clustersetIPEnabledKey]), &clustersetIPInfo.Enabled)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error reading clusersetIPEnabled status")
+	}
+
+	err = json.Unmarshal([]byte(configMap.Data[clustersetIPClusterSize]), &clustersetIPInfo.AllocationSize)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error reading ClustersetIPClusterSize")
+	}
+
+	err = json.Unmarshal([]byte(configMap.Data[clustersetIPCidrRange]), &clustersetIPInfo.CIDR)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error reading ClustersetIPCidrRange")
+	}
+
+	clustersetIPInfo.Clusters, err = cidr.ExtractClusterInfo(configMap)
+
+	return &clustersetIPInfo, configMap, err //nolint:wrapcheck // No need to wrap
+}
+
+func AssignClustersetIPs(clustersetIPInfo *Info, netconfig Config, status reporter.Interface) (string, error) {
```

Comment:
This doesn't need to be public.

```suggestion
func assignClustersetIPs(clustersetIPInfo *Info, netconfig Config, status reporter.Interface) (string, error) {
```

Reply:
done